stellargraph/layer/watch_your_step.py

```diff
@@ -0,0 +1,199 @@
+            regularizer=self.attention_regularizer,
+            constraint=self.attention_constraint,
+        )
+
+        self.built = True
+
+    def call(self, partial_powers):
+        """
+        Applies the layer and calculates the expected random walks.
+
+        Args:
+            partial_powers: num_rows rows of the first num_powers powers of adjacency matrix with shape
+            (num_rows, num_powers, num_nodes)
+
+        Returns:
+            Tensor that represents the expected random walks starting from nodes corresponding to the input rows of
+            shape (num_rows, num_nodes)
+        """
+
+        attention = K.softmax(self.attention_weights)
+        expected_walk = tf.einsum("ijk,j->ik", partial_powers, attention)
+
+        return expected_walk
+
+
+@experimental(reason="lack of unit tests")
+class WatchYourStep:
+    """
+    Implementation of the node embeddings as in Watch Your Step: Learning Node Embeddings via Graph Attention
+    https://arxiv.org/pdf/1710.09599.pdf.
+
+    This model requires specification of the number of random walks starting from each node, and the embedding dimension
+    to use for the node embeddings. Note, that the embedding dimension should be an even number or else will be
+    rounded down to nearest even number.
+
+    .. warning::
+
+        This class is experimental: it is insufficiently tested.
+
+    Args:
+        generator (AdjacencyPowerGenerator): the generator
+        num_walks (int): the number of random walks starting at each node to use when calculating the expected random
+        walks.
+        embedding dimension (int): the dimension to use for the node embeddings
+    """
+
+    def __init__(
+        self, generator, num_walks, embedding_dimension, attention_regularizer=None
+    ):
+
+        if not isinstance(generator, AdjacencyPowerGenerator):
+            raise TypeError(
+                "generator should be an instance of AdjacencyPowerGenerator."
+            )
+
+        if not isinstance(num_walks, int):
+            raise TypeError("num_walks should be an int.")
+
+        if num_walks <= 0:
+            raise ValueError("num_walks should be a positive int.")
+
+        self.num_walks = num_walks
+        self.num_powers = generator.num_powers
+        self.n_nodes = int(generator.Aadj_T.shape[0])
+        self.embedding_dimension = embedding_dimension
+        self.attention_regularizer = attention_regularizer
+
+    def build(self):
+        """
+        This function builds the layers for a keras model.
+
+        returns:
+            A tuple of (inputs, outputs) to use with a keras model.
+        """
+
+        input_rows = Input(batch_shape=(None,), name="row_node_ids", dtype="int64")
+        input_powers = Input(batch_shape=(None, self.num_powers, self.n_nodes))
+
+        left_embedding = Embedding(
+            self.n_nodes,
+            self.embedding_dimension,
+            input_length=None,
+            name="LEFT_EMBEDDINGS",
```

Comment:
Is it worth writing, just to avoid any ambiguities with other parts of a large model:

```suggestion
            name="WATCH_YOUR_STEP_LEFT_EMBEDDINGS",
```

Reply:
good idea!